File: Python projects/Python Fundamentals/11_final_exam/02._Programming_Fundamentals_Final_Exam/01_world_tour.py
tour_stops = input()
cmd = [el for el in input().split(":")]

while not cmd[0] == 'Travel':
    task = cmd[0]
    if task == 'Add Stop':
        index, substr = int(cmd[1]), cmd[2]
        # first check if index is valid
        if index in range(len(tour_stops)):
            tour_stops = tour_stops[:index] + substr + tour_stops[index:]
        print(f"{tour_stops}")
    elif task == 'Remove Stop':
        start_index, end_index = int(cmd[1]), int(cmd[2])
        if start_index in range(len(tour_stops)) and end_index in range(len(tour_stops)):
            tour_stops = tour_stops[:start_index] + tour_stops[end_index + 1:]
        print(f"{tour_stops}")
    elif task == 'Switch':
        old_string, new_string = cmd[1], cmd[2]
        if old_string in tour_stops:
            tour_stops = tour_stops.replace(old_string, new_string)
        print(f"{tour_stops}")
    cmd = [el for el in input().split(":")]
print(f"Ready for world tour! Planned stops: {tour_stops}")


'''
TASK:
On the first line, you will be given a string containing all of your stops. Until you receive the command "Travel", you 
will be given some commands to manipulate that initial string. The commands can be:
"Add Stop:{index}:{string}":
Insert the given string at that index only if the index is valid.
"Remove Stop:{start_index}:{end_index}":
Remove the elements of the string from the starting index to the end index (inclusive) if both indices are valid.
"Switch:{old_string}:{new_string}":
If the old string is in the initial string, replace it with the new one (all occurrences).
Note: After each command, print the current state of the string!
After the "Travel" command, print the following: "Ready for world tour! Planned stops: {string}".
Input / Constraints
JavaScript: you will receive a list of strings.
An index is valid if it is between the first and the last element index (inclusive) (0 ….. Nth) in the sequence.
Output
Print the proper output messages in the proper cases as described in the problem description.
'''

File: JavaScript projects/JavaScript Fundamentals/Final Exam/Exam_02/01_world_tour.js
function solve(input){
    let route = input.shift();
    let cmd = input.shift();

    while (cmd !== 'Travel'){
        cmd = cmd.split(':');
        let action = cmd.shift();

        if (action === 'Add Stop'){
            let index = Number(cmd.shift());
            let stop = cmd.shift();
            
            if (0 <= index && index < route.length){
            route = route.slice(0, index) + stop + route.slice(index , route.length);
            }

        }else if (action === 'Remove Stop'){
            let start = Number(cmd.shift());
            let end = Number(cmd.shift());

             if ((0 <= start && start < route.length) && 0 <= end && end < route.length){
            route = route.slice(0, start) + route.slice(end + 1, route.length);
            }

        }else if (action === 'Switch'){
            let oldString = cmd.shift();
            let newString = cmd.shift();
            
            if (route.includes(oldString)){
                route = route.replace(oldString, newString);
            }
        }
        console.log(`${route}`);

        cmd = input.shift();
    }
    console.log(`Ready for world tour! Planned stops: ${route}`);
}

solve(["Hawai::Cyprys-Greece",
"Add Stop:7:Rome",
"Remove Stop:11:16",
"Switch:Hawai:Bulgaria",
"Travel"]);
solve(["Albania:Bulgaria:Cyprus:Deuchland",
"Add Stop:3:Nigeria",
"Remove Stop:4:8",
"Switch:Albania: Azərbaycan",
"Travel"]);
